developers_chamber/qa/base.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import subprocess
+
+import click
+from git import Repo
+
+
+LOGGER = logging.getLogger()
+
+
+class QAError(Exception):
+    """
+    Generic QA exception that also carries command output.
+    """
+    def __init__(self, msg, output):
+        super().__init__(msg)
+        self.output = output.strip()
+
+
+class QACheck:
+    """
+    Base class for a quality assurance check.
+
+    Arguments:
+        name: Name of the check.
+    """
+    name = None
+
+    def _get_repo(self):
+        """
+        Returns the repo object.
+        """
+        return Repo('.')
+
+    def _get_default_branch(self):
+        """
+        Returns default branch of the repo.
+        """
+        return self._run_command("git symbolic-ref refs/remotes/origin/HEAD | sed 's@^refs/remotes/origin/@@'")
+
+    def _get_diffs(self):
+        """
+        Returns Git diffs against default branch.
+        """
+        repo = self._get_repo()
+        target_branch = getattr(repo.remotes.origin.refs, self._get_default_branch())
+        target_commit = repo.merge_base(repo.active_branch, target_branch)[0]
+        return target_commit.diff(repo.active_branch.name, create_patch=True)
+
+    def _get_unstaged(self):
+        """
+        Returns unstaged files in the repo.
+        """
+        return self._get_repo().index.diff(None, create_patch=True)
+
+    def _run_command(self, command):
+        """
+        Runs shell command and returns its output.
+        """
+        try:
+            return subprocess.check_output(command, shell=True).decode().strip()
+        except subprocess.CalledProcessError as ex:
+            raise QAError(str(ex), ex.output.decode())
+
+    def _run_check(self):
+        """
+        This function should implement the actual check and should raise `QAError` if failed.
+        """
+        raise NotImplementedError()
+
+    def _cleanup(self):
+        """
+        Cleans up the repo to be fresh for another check.
+        """
+        self._get_repo().git.stash('save')
+
+    def run(self):
+        """
+        Runs the check and cleanup methods.
+        """
+        try:
+            self._run_check()
+        finally:
+            self._cleanup()
+
+
+class QACheckRunner:
+    """
+    Runs multiple checks and evaluates the results while printing a nice output.
+    """
+    LINE_LENGTH = 60
```

Comment:
Why 60?

Reply:
Arbitrary choice.

Reply:
I'd make it at least 80 then. Out of habit.

Reply:
Yes 80 is standard value (for us it is 120, but for cicle output 80 will be better)